File: Csharp/SolidCourse/ISP/Before.cs
// Large interface
public interface IMachine
{
    void Print(string document);
    void Fax(string document);
}

// Printer class that doesn't need faxing capability
public class Printer : IMachine
{
    public void Print(string document)
    {
        Console.WriteLine($"Printing: {document}");
    }

    public void Fax(string document)
    {
        // Printer doesn't implement faxing, so we throw an exception
        throw new NotSupportedException("Faxing not supported.");
    }
}

// Printer class that doesn't need faxing capability
public class FaxMachine : IMachine
{
    public void Print(string document)
    {
        throw new NotSupportedException("Printing not supported.");
    }

    public void Fax(string document)
    {
        // Printer doesn't implement faxing, so we throw an exception
        throw new NotSupportedException("Faxing not supported.");
    }
}
// Client code
public class Office
{
    public void ProcessDocument(IMachine machine, string document)
    {
        machine.Print(document);
        machine.Fax(document); // This would cause an issue if machine is a Printer
    }
}

File: CPlusPlus/ISP/Before.cpp
// Machine.cpp
#include <iostream>
#include <stdexcept>
#include <string>

class IMachine {
public:
    virtual void Print(const std::string& document) = 0;
    virtual void Fax(const std::string& document) = 0;
    virtual ~IMachine() = default;
};

// Printer class forced to implement Fax
class Printer : public IMachine {
public:
    void Print(const std::string& document) override {
        std::cout << "Printing: " << document << std::endl;
    }

    void Fax(const std::string& document) override {
        throw std::logic_error("Faxing not supported.");
    }
};

// Fax machine forced to implement Print
class FaxMachine : public IMachine {
public:
    void Print(const std::string& document) override {
        throw std::logic_error("Printing not supported.");
    }

    void Fax(const std::string& document) override {
        throw std::logic_error("Faxing not supported.");
    }
};

// Client code
class Office {
public:
    void ProcessDocument(IMachine& machine, const std::string& document) {
        machine.Print(document);
        machine.Fax(document); // Might throw depending on implementation
    }
};
